### DISCUSSION TITLE
Protecting against incorrect string subscription

### DISCUSSION THREAD

**Discussion by Apakottur** [python/typing#1481] [General]
Is there any typing trick that can protect against the following coding mistake:

```python
fruits: list[str] = ["banana", "apple"]
first_fruit: str = my_list[0][0]
print(first_fruit) # Prints "b" and not "banana"
```

I'm working with a huge codebase where we are doing a big refactor where [0][0] is being changed to [0] due to an API change, but since str[0] is still a string, everything type checks, even if we forget to remove the double subscription.
We almost never want to actually get a single character from a string, and already had two major production bugs due to forgotten double subscripts.

If there's something that mypy can catch - amazing, but a runtime error would be also be valuable, it's better than the code continuing to execute with the incorrect string.

I was thinking in the direction of creating a subclass of `str` and casting/typing fruits to `list[MyStr]` and overriding the `getitem` function to return an error, but was wondering if there's something better.

**Comment by erictraut** [comment]
You could temporarily comment out the `str.__getitem__` method in `builtins.pyi`, then run the type checker. This wouldn't be a permanent solution, but it could help you get through your big refactor.
